dc_logging_client/log_entries.py

```diff
@@ -37,17 +37,25 @@ def __post_init__(self):
 
 
 @dataclass
-class DummyLogEntry(BaseLogEntry, ValidDCProductMixin):
-    text: str
+class UTMMixin:
+    utm_source: Union[None, str] = field(default_factory=str)
+    utm_campaign: Union[None, str] = field(default_factory=str)
 
 
 @dataclass
-class PostcodeLogEntry(BaseLogEntry, ValidDCProductMixin):
-    postcode: str
+class DummyLogEntry(BaseLogEntry, UTMMixin, ValidDCProductMixin):
+    text: str = field(default_factory=str)
+
+
+@dataclass
+class PostcodeLogEntry(BaseLogEntry, UTMMixin, ValidDCProductMixin):
+    postcode: str = field(default_factory=str)
     timestamp: Union[datetime.datetime, str] = ""
     api_key: str = ""
 
     def __post_init__(self):
         super().__post_init__()
+        if not self.postcode:
```

Comment:
There may be a reason elsewhere you want to raise a `ValueError` but removing the default on the field definition would enforce the need for a postcode without having to add this

Reply:
Yes, you are right, however I'm also keen on making other args optional.

I can't quite figure this out, but dataclasses turn the fields in to `_init__` args on the returned class, and it's easy to end up in a situation where some args are required and some aren't but they get defined in the wrong order.

For example, when I remove the field I get 

` TypeError: non-default argument 'postcode' follows default argument`

I think this is caused by the MRO and mixins: the fields from the `ValidDCProductMixin` and `UTMMixin` are added to the class _before_ the fields defined on the class.

If you can think of a better way around this I'd like to know about it. I could dump the mixins and just add to the class, but it feels like this is a useful pattern.

Reply:
Oh that is interesting. Looks like that is this issue https://bugs.python.org/issue39300. So it seems the fields from the mixins are inherited and added to the `__init__` method first. Cant think of a good way around of it without removing the mixins, but I agree its nice to have the mixins so I'd leave as is.